src/modules/ims_usrloc_pcscf/udomain.c

```diff
@@ -1077,91 +1113,408 @@ int preload_udomain(db1_con_t* _c, udomain_t* _d)
+        char *p_pua = "DELETE from pua where pua.pres_uri =\"%.*s\"";
+        char *pua = "pua";
+
+        str query_pua, query_buffer;
+        query_pua.s = p_pua;
+        query_pua.len = strlen(query_pua.s);
+
+        _pua.s = pua;
+        _pua.len = strlen(_pua.s);
+
+        db_handle = get_db_handle ();
+        if (db_handle == 0){
+           LM_ERR("failed to get valid db_handle");
+           return 0;
+        }
+   
+
+        if (ul_dbf.use_table(db_handle, &_pua) < 0) {
+                LM_ERR("failed to use table pua");
+                return 0;
+        }
+
+        int len = query_pua.len + presentity_uri->len + 2;
+        query_buffer.s = (char*) pkg_malloc(len);
+        if (!query_buffer.s) {
+                LM_ERR("mo more pkg mem\n");
+                return 0;
+        }
+        snprintf(query_buffer.s, len, p_pua, presentity_uri->len, presentity_uri->s);
+        query_buffer.len = strlen(query_buffer.s);//len;
+
+
+        LM_INFO("QUERY IS [%.*s] and len is %d\n", query_buffer.len, query_buffer.s, query_buffer.len);
+
+        if (ul_dbf.raw_query(db_handle, &query_buffer, &res) != 0) {
+             LM_ERR("Unable to delete pua entry for   [%.*s]\n",
+                                        presentity_uri->len, presentity_uri->s);
+             ul_dbf.free_result(db_handle, res);
+             if (query_buffer.s)
+                        pkg_free(query_buffer.s);
+             return 0;
+        }
+
+        ul_dbf.free_result(db_handle, res);
+        if (query_buffer.s)
+                        pkg_free(query_buffer.s);
+        return 1;
+}
+
+int db_load_pcontact(udomain_t* _d, str *_aor, int insert_cache, struct pcontact** _c, pcontact_info_t* contact_info)
 {
 	pcontact_info_t *ci;
-	db_key_t columns[15];
-	db_key_t keys[1];
-	db_val_t vals[1];
+    pcontact_t* c=NULL; //Modified
 	db1_res_t* res = NULL;
 	db_row_t *row;
-	int i;
-	str aor;
-
-	pcontact_t* c;
-
-	keys[0] = &aor_col;
-	vals[0].type = DB1_STR;
-	vals[0].nul = 0;
-	vals[0].val.str_val = *_aor;
-
-	columns[0] = &domain_col;
-	columns[1] = &aor_col;
-	columns[2] = &host_col;
-	columns[3] = &port_col;
-	columns[4] = &protocol_col;
-	columns[5] = &received_col;
-	columns[6] = &received_port_col;
-	columns[7] = &received_proto_col;
-	columns[8] = &rx_session_id_col;
-	columns[9] = &reg_state_col;
-	columns[10] = &expires_col;
-	columns[11] = &socket_col;
-	columns[12] = &service_routes_col;
-	columns[13] = &public_ids_col;
-	columns[14] = &path_col;
+	int i, len, querylen;
+        str aor, query_location, query_buffer, port={0,0};
+        db1_con_t* db_handle = 0; 
+        char *p_location;
+
+        if (_aor->len>0 && _aor->s){
+                LM_INFO("Querying database for P-CSCF contact [%.*s]\n", _aor->len, _aor->s);
+
+    p_location = "SELECT  domain,aor,host,port,protocol,received,received_port,received_proto,rx_session_id,reg_state,expires,socket,service_routes,public_ids,path from location WHERE aor=\"%.*s\""; // AND reg_state=%d";
```

Comment:
This is not the way you should interact with the DB API. There are actually methods to build a select query. You can find some examples e.g. in the usrloc/dlist.c or in usrloc_db.c files.
This way the module will work with all database drivers that implement the proper functions.